datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,829 @@
+        let ids = ids.as_any().downcast_ref::<StringArray>().ok_or_else(|| {
+            DataFusionError::Execution("Expected StringArray".to_string())
+        })?;
+        let vals = vals.as_primitive::<VAL>();
+        let null_count = vals.null_count();
+        for row_idx in 0..ids.len() {
+            if null_count > 0 && vals.is_null(row_idx) {
+                continue;
+            }
+            let val = vals.value(row_idx);
+            let id = if ids.is_null(row_idx) {
+                None
+            } else {
+                // Check goes here, because it is generalizable between str/String and Row/OwnedRow
+                let id = ids.value(row_idx);
+                if self.priority_map.is_full() {
+                    if self.priority_map.desc {
+                        if let Some(worst) = self.priority_map.min_val() {
+                            if val < *worst {
+                                continue;
+                            }
+                        }
+                    } else if let Some(worst) = self.priority_map.max_val() {
+                        if val > *worst {
+                            continue;
+                        }
+                    }
+                }
+                Some(id.to_string())
+            };
+
+            self.priority_map.insert(id, val)?;
+        }
+        Ok(())
+    }
+
+    fn emit(&mut self) -> Result<Vec<ArrayRef>> {
+        let (keys, vals): (Vec<_>, Vec<_>) =
+            self.priority_map.drain().into_iter().unzip();
+        let keys = Arc::new(StringArray::from(keys));
+        let vals = Arc::new(PrimitiveArray::<VAL>::from_iter_values(vals));
+        Ok(vec![keys, vals])
+    }
+
+    fn is_empty(&self) -> bool {
+        self.priority_map.is_empty()
+    }
+}
+
+impl Stream for GroupedTopKAggregateStream {
+    type Item = Result<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        while let Poll::Ready(res) = self.input.poll_next_unpin(cx) {
+            match res {
+                // got a batch, convert to rows and append to our TreeMap
+                Some(Ok(batch)) => {
+                    self.started = true;
+                    trace!(
+                        "partition {} has {} rows and got batch with {} rows",
+                        self.partition,
+                        self.row_count,
+                        batch.num_rows()
+                    );
+                    if log::log_enabled!(Level::Trace) && batch.num_rows() < 20 {
+                        print_batches(&[batch.clone()])?;
+                    }
+                    self.row_count += batch.num_rows();
+                    let batches = &[batch];
+                    let group_by_values =
+                        evaluate_group_by(&self.group_by, batches.first().unwrap())?;
+                    let group_by_values =
```

Comment:
maybe it would be clearer if you asserted that these slices were length 1 and then do group_by_values[0][0] or soemthing